### PR TITLE
feat(backend): start of handling grant token rotation more gracefully

### PR DESCRIPTION
**Pull request by mkurapov**
## Changes proposed in this pull request

- Adding `getOrCreate` and `delete` methods to `backend` grant service:
    - `getOrCreate` method:
        - soft deletes a grant if token rotation fails
        - uses the open payments client to request grant and store it instead of having it be the responsibility of the caller
    - `delete` method
        - allows callers to soft delete a grant if necessary
    - **these methods are still unused at this point, I will be updating callers to use them in a later PR, so we can break up the changes**
- Adding `deletedAt` field to the grants table to allow soft deletion, as well as dropping the unique constraint on `authServerId` `accessType` and `accessActions` such that we can properly soft delete problematic grants 
       
    
       
## Context

- This is the first part of #2852:
    - We are making the grant service more resilient to edge cases (like being unable to rotate tokens if they are missing from the receivers Rafiki DB)

## Checklist

- [x] Related issues linked using `fixes #number`
- [x] Tests added/updated
- [ ] Documentation added
- [ ] Make sure that all checks pass
- [ ] Bruno collection updated

### CODE REVIEW

File: packages/backend/migrations/20240820101148_drop_unique_grants.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+exports.up = function (knex) {
+  return knex.schema.alterTable('grants', function (table) {
+    table.dropUnique(['authServerId', 'accessType', 'accessActions'])
```
Comment (mkurapov) [inline packages/backend/migrations/20240820101148_drop_unique_grants.js:7]: Dropping the unique constraint on authServerId accessType and accessActions such that we can properly soft delete grants (see migration below)

File: packages/backend/src/app.ts [hunk 2]
```diff
@@ -232,6 +234,8 @@ export interface AppServices {
   incomingPaymentService: Promise<IncomingPaymentService>
   remoteIncomingPaymentService: Promise<RemoteIncomingPaymentService>
   receiverService: Promise<ReceiverService>
+  grantService: Promise<GrantService>
+  authServerService: Promise<AuthServerService>
```
Comment (mkurapov) [inline packages/backend/src/app.ts:238]: weren't included before, adding here so we get proper types

File: packages/backend/src/open_payments/grant/service.ts [hunk 3]
```diff
@@ -1,16 +1,26 @@
...
+  isPendingGrant
+} from '@interledger/open-payments'
+import { GrantError } from './errors'
 
 export interface GrantService {
   create(options: CreateOptions): Promise<Grant>
   get(options: GrantOptions): Promise<Grant | undefined>
   update(grant: Grant, options: UpdateOptions): Promise<Grant>
+  getOrCreate(options: GrantOptions): Promise<Grant | GrantError>
+  delete(id: string): Promise<Grant>
```
Comment (mkurapov) [inline packages/backend/src/open_payments/grant/service.ts:18]: still unused

File: packages/backend/src/open_payments/grant/service.ts [hunk 4]
```diff
@@ -89,6 +101,140 @@ async function updateGrant(
...
+  deps: ServiceDependencies,
+  options: GrantOptions
+): Promise<Grant | undefined> {
+  return await Grant.query(deps.knex)
+    .findOne({
+      accessType: options.accessType
+    })
+    .whereNull('deletedAt')
+    .andWhere('authServer.url', options.authServer)
+    .andWhere('accessActions', '@>', options.accessActions) // all options.accessActions are a subset of saved accessActions
```
Comment (mkurapov) [inline packages/backend/src/open_payments/grant/service.ts:134]: e.g. if we have `[AccessActions.Create, AccessActions.Complete, AccessActions.Read]` and we are trying to getOrCreate a grant with `[AccessActions.Create]` we will properly return it, but if one of the elements in the requested list doesnt exist, we will not be able to fetch it

File: packages/backend/src/app.ts (hunk 2 above)
Comment (BlairCurrey) [reply packages/backend/src/app.ts:238]: easy to miss this

File: packages/backend/src/open_payments/grant/service.ts (hunk 4 above)
Comment (BlairCurrey) [reply packages/backend/src/open_payments/grant/service.ts:134]: It looks like this differs slightly from the `getGrant` which finds one that has strictly equal `accessActions`.  Its not quite clear what the ramifications of this will be. Can you speak to this difference?

File: packages/backend/migrations/20240820101148_drop_unique_grants.js (hunk 1 above)
Comment (BlairCurrey) [reply packages/backend/migrations/20240820101148_drop_unique_grants.js:7]: seems similar to this case? https://github.com/interledger/rafiki/pull/2863#discussion_r1720065804

So could we use a partial unique index (where `deletedAt` is not null)?

File: packages/backend/src/open_payments/grant/service.ts (hunk 4 above)
Comment (mkurapov) [reply packages/backend/src/open_payments/grant/service.ts:134]: before, if we already had a grant for `[Read, Create]` and then we were trying to getOrCreate a grant with `[Read]`, it would fail to get the existing one, even though the existing grant has the sufficient access requirements

Comment (BlairCurrey) [reply packages/backend/src/open_payments/grant/service.ts:134]: I suppose we're getting rid of the `getGrant` here then (in subsequent PR)? looks like its unused but if we keep it for some reason I guess we should update to filter out by `deletedAt` (and maybe this same `accessActions` where clause). But I imagine we can just remove it.

Comment (mkurapov) [reply packages/backend/src/open_payments/grant/service.ts:134]: Yes exactly, will be removing the unused methods

File: packages/backend/migrations/20240820101148_drop_unique_grants.js (hunk 1 above)
Comment (mkurapov) [reply packages/backend/migrations/20240820101148_drop_unique_grants.js:7]: I don't think this unique constraint is very useful to begin with, particularly because order matters in `accessActions` array, meaning [Create, Read] is different from [Read, Create]

File: packages/backend/migrations/20240820101201_add_deleted_at_grants.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+/**
```
Comment (njlie) [inline packages/backend/migrations/20240820101201_add_deleted_at_grants.js:1]: nit: why not combine this migration with `drop_unique_grants`?

Comment (mkurapov) [reply packages/backend/migrations/20240820101201_add_deleted_at_grants.js:1]: Yup, makes sense, let me do that

File: packages/backend/src/open_payments/grant/service.test.ts [hunk 6]
```diff
@@ -201,4 +219,400 @@ describe('Grant Service', (): void => {
...
+        accessType: AccessType.IncomingPayment,
+        accessActions: [AccessAction.ReadAll]
+      }
+
+      await expect(
+        getExistingGrant({ knex } as ServiceDependencies, options)
+      ).resolves.toEqual({ ...existingGrant, authServer })
+    })
+
+    test('gets existing grant (requested actions are a subset of saved actions)', async () => {
```
Comment (njlie) [inline packages/backend/src/open_payments/grant/service.test.ts:485]: Would be necessary to create another test (or modify this one) to ensure that it can retrieve existing grants that have (Read/List)-All rights when (Read/List) is requested?

Comment (mkurapov) [reply packages/backend/src/open_payments/grant/service.test.ts:485]: good point, maybe during create if ReadAll/ListAll is requested, we can also manually add the corresponding Read/List as well

**Review by njlie — APPROVED** [review]
LGTM
